Clean up "pattern doesn't bind x" messages

estebank opened the pull request:
Group "missing variable bind" spans in `or` matches and clarify wording
for the two possible cases: when a variable from the first pattern is
not in any of the subsequent patterns, and when a variable in any of the
other patterns is not in the first one.

Before:

```rust
error[E0408]: variable `a` from pattern #1 is not bound in pattern #2
  --> file.rs:10:23
   |
10 |         T::T1(a, d) | T::T2(d, b) | T::T3(c) | T::T4(a) => { println!("{:?}", a); }
   |                       ^^^^^^^^^^^ pattern doesn't bind `a`

error[E0408]: variable `b` from pattern #2 is not bound in pattern #1
  --> file.rs:10:32
   |
10 |         T::T1(a, d) | T::T2(d, b) | T::T3(c) | T::T4(a) => { println!("{:?}", a); }
   |                                ^ pattern doesn't bind `b`

error[E0408]: variable `a` from pattern #1 is not bound in pattern #3
  --> file.rs:10:37
   |
10 |         T::T1(a, d) | T::T2(d, b) | T::T3(c) | T::T4(a) => { println!("{:?}", a); }
   |                                     ^^^^^^^^ pattern doesn't bind `a`

error[E0408]: variable `d` from pattern #1 is not bound in pattern #3
  --> file.rs:10:37
   |
10 |         T::T1(a, d) | T::T2(d, b) | T::T3(c) | T::T4(a) => { println!("{:?}", a); }
   |                                     ^^^^^^^^ pattern doesn't bind `d`

error[E0408]: variable `c` from pattern #3 is not bound in pattern #1
  --> file.rs:10:43
   |
10 |         T::T1(a, d) | T::T2(d, b) | T::T3(c) | T::T4(a) => { println!("{:?}", a); }
   |                                           ^ pattern doesn't bind `c`

error[E0408]: variable `d` from pattern #1 is not bound in pattern #4
  --> file.rs:10:48
   |
10 |         T::T1(a, d) | T::T2(d, b) | T::T3(c) | T::T4(a) => { println!("{:?}", a); }
   |                                                ^^^^^^^^ pattern doesn't bind `d`

error: aborting due to 6 previous errors
```

After:

```rust
error[E0408]: variable `d` is not bound in all patterns
  --> $DIR/issue-39698.rs:20:37
   |
20 |         T::T1(a, d) | T::T2(d, b) | T::T3(c) | T::T4(a) => { println!("{:?}", a); }
   |                  -          -       ^^^^^^^^   ^^^^^^^^ pattern doesn't bind `d`
   |                  |          |       |
   |                  |          |       pattern doesn't bind `d`
   |                  |          variable not in all patterns
   |                  variable not in all patterns

error[E0408]: variable `c` is not bound in all patterns
  --> $DIR/issue-39698.rs:20:48
   |
20 |         T::T1(a, d) | T::T2(d, b) | T::T3(c) | T::T4(a) => { println!("{:?}", a); }
   |         ^^^^^^^^^^^   ^^^^^^^^^^^         -    ^^^^^^^^ pattern doesn't bind `c`
   |         |             |                   |
   |         |             |                   variable not in all patterns
   |         |             pattern doesn't bind `c`
   |         pattern doesn't bind `c`

error[E0408]: variable `a` is not bound in all patterns
  --> $DIR/issue-39698.rs:20:37
   |
20 |         T::T1(a, d) | T::T2(d, b) | T::T3(c) | T::T4(a) => { println!("{:?}", a); }
   |               -       ^^^^^^^^^^^   ^^^^^^^^         - variable not in all patterns
   |               |       |             |
   |               |       |             pattern doesn't bind `a`
   |               |       pattern doesn't bind `a`
   |               variable not in all patterns

error[E0408]: variable `b` is not bound in all patterns
  --> $DIR/issue-39698.rs:20:37
   |
20 |         T::T1(a, d) | T::T2(d, b) | T::T3(c) | T::T4(a) => { println!("{:?}", a); }
   |         ^^^^^^^^^^^            -    ^^^^^^^^   ^^^^^^^^ pattern doesn't bind `b`
   |         |                      |    |
   |         |                      |    pattern doesn't bind `b`
   |         |                      variable not in all patterns
   |         pattern doesn't bind `b`

error: aborting due to 4 previous errors
```

Fixes #39698.

estebank committed:
Clean up "pattern doesn't bind x" messages

Group "missing variable bind" spans in `or` matches and clarify wording
for the two possible cases: when a variable from the first pattern is
not in any of the subsequent patterns, and when a variable in any of the
other patterns is not in the first one.

Before:

```
error[E0408]: variable `a` from pattern #1 is not bound in pattern #2
  --> file.rs:10:23
   |
10 |         T::T1(a, d) | T::T2(d, b) | T::T3(c) | T::T4(a) => { println!("{:?}", a); }
   |                       ^^^^^^^^^^^ pattern doesn't bind `a`

error[E0408]: variable `b` from pattern #2 is not bound in pattern #1
  --> file.rs:10:32
   |
10 |         T::T1(a, d) | T::T2(d, b) | T::T3(c) | T::T4(a) => { println!("{:?}", a); }
   |                                ^ pattern doesn't bind `b`

error[E0408]: variable `a` from pattern #1 is not bound in pattern #3
  --> file.rs:10:37
   |
10 |         T::T1(a, d) | T::T2(d, b) | T::T3(c) | T::T4(a) => { println!("{:?}", a); }
   |                                     ^^^^^^^^ pattern doesn't bind `a`

error[E0408]: variable `d` from pattern #1 is not bound in pattern #3
  --> file.rs:10:37
   |
10 |         T::T1(a, d) | T::T2(d, b) | T::T3(c) | T::T4(a) => { println!("{:?}", a); }
   |                                     ^^^^^^^^ pattern doesn't bind `d`

error[E0408]: variable `c` from pattern #3 is not bound in pattern #1
  --> file.rs:10:43
   |
10 |         T::T1(a, d) | T::T2(d, b) | T::T3(c) | T::T4(a) => { println!("{:?}", a); }
   |                                           ^ pattern doesn't bind `c`

error[E0408]: variable `d` from pattern #1 is not bound in pattern #4
  --> file.rs:10:48
   |
10 |         T::T1(a, d) | T::T2(d, b) | T::T3(c) | T::T4(a) => { println!("{:?}", a); }
   |                                                ^^^^^^^^ pattern doesn't bind `d`

error: aborting due to 6 previous errors
```

After:

```
error[E0408]: variable `a` is not bound in all patterns
  --> file.rs:20:37
   |
20 |         T::T1(a, d) | T::T2(d, b) | T::T3(c) | T::T4(a) => {
intln!("{:?}", a); }
   |               -       ^^^^^^^^^^^   ^^^^^^^^         - variable
t in all patterns
   |               |       |             |
   |               |       |             pattern doesn't bind `a`
   |               |       pattern doesn't bind `a`
   |               variable not in all patterns

error[E0408]: variable `d` is not bound in all patterns
  --> file.rs:20:37
   |
20 |         T::T1(a, d) | T::T2(d, b) | T::T3(c) | T::T4(a) => {
intln!("{:?}", a); }
   |                  -          -       ^^^^^^^^   ^^^^^^^^ pattern
esn't bind `d`
   |                  |          |       |
   |                  |          |       pattern doesn't bind `d`
   |                  |          variable not in all patterns
   |                  variable not in all patterns

error[E0408]: variable `b` is not bound in all patterns
  --> file.rs:20:37
   |
20 |         T::T1(a, d) | T::T2(d, b) | T::T3(c) | T::T4(a) => {
intln!("{:?}", a); }
   |         ^^^^^^^^^^^            -    ^^^^^^^^   ^^^^^^^^ pattern
esn't bind `b`
   |         |                      |    |
   |         |                      |    pattern doesn't bind `b`
   |         |                      variable not in all patterns
   |         pattern doesn't bind `b`

error[E0408]: variable `c` is not bound in all patterns
  --> file.rs:20:48
   |
20 |         T::T1(a, d) | T::T2(d, b) | T::T3(c) | T::T4(a) => {
intln!("{:?}", a); }
   |         ^^^^^^^^^^^   ^^^^^^^^^^^         -    ^^^^^^^^ pattern
esn't bind `c`
   |         |             |                   |
   |         |             |                   variable not in all
tterns
   |         |             pattern doesn't bind `c`
   |         pattern doesn't bind `c`

error: aborting due to 4 previous errors
```

* Have only one presentation for binding consistency errors
* Point to same binding in multiple patterns when possible
* Check inconsistent bindings in all arms
* Simplify wording of diagnostic message
* Sort emition and spans of binding errors for deterministic output

rust-highfive commented:
r? @eddyb

(rust_highfive has picked a reviewer for you, use r? to override)

estebank commented:
Colored output:

<img width="717" alt="" src="https://cloud.githubusercontent.com/assets/1606434/22811074/86db3784-eef1-11e6-94f7-c633bac45451.png">

osa1 commented:
This is still inconsistent I think. I don't understand why the first error message is any different than the second one. If first one reports that a is missing from the second and third patterns, second shoyuld return that b is missing from first, third, and fourth pattern _in the exact same way_. I don't see why would you say the same thing in a different way second time.

eddyb commented:
r? @jonathandturner

jonathandturner commented:
@estebank - any thoughts on the proposal at https://github.com/rust-lang/rust/issues/39698#issuecomment-279244282 ?

estebank commented:
@jonathandturner the current output looks this way, which I believe is what you were proposing in that comment:

<img width="635" alt="" src="https://cloud.githubusercontent.com/assets/1606434/22950130/77923dc4-f2b9-11e6-9d41-48e8834e9160.png">

solson commented:
It's a minor thing, but do you think it's worth sorting the pattern numbers?

Instead of: `isn't bound in patterns #2, #4, #1`

We could have: `isn't bound in pattern #1, #2, #4`

solson commented:
In the example in the screenshot, can we avoid printing a message for variable `a` twice? The first two errors seem redundant.

estebank commented:
@solson:

<img width="640" alt="" src="https://cloud.githubusercontent.com/assets/1606434/22953160/e25a8dcc-f2c4-11e6-86be-56cc1f237ef5.png">

solson commented:
@estebank Looking good! There are some errors in the output now, though. The first error's text talks about variable `c` but its spans are pointing to `d` and patterns missing `d`, for example.

jonathandturner commented:
@estebank - is that latest screenshot the real thing?  If so, the second error looks wrong.  Shouldn't it be 'c'?

solson commented:
They're actually all wrong. :) There's a mismatch between the text and spans.

estebank commented:
@solson @jonathandturner [copy paste error](https://github.com/rust-lang/rust/pull/39713/commits/4e43853f78af8dac580916c652ca86c806e9cd72)

solson commented:
@estebank Looks great!

My only remaining question is whether we should have `variable not in all patterns` for every instance of the binding, but I don't mind how it is now.

jonathandturner commented:
Yeah, looks good.  Once we get the travis stuff sorted will review.  Is that specific to this PR or a general travis issue?

estebank commented:
> My only remaining question is whether we should have variable not in all patterns for every instance of the binding or just the first, but I don't mind how it is now.

I'd lean on keeping the label.

> Yeah, looks good. Once we get the travis stuff sorted will review. Is that specific to this PR or a general travis issue?

It's a regression:

```
  0 - |error[E0409]: variable `y` is bound with different mode in pattern #2 than in pattern #1|
    + |error[E0409]: variable `y` is bound with different mode in pattern #1 than in pattern #1|
```

[I just fixed it](https://github.com/rust-lang/rust/pull/39713/commits/39de12f6f492047417ce703d10d901a4325dec10).

jonathandturner commented:
Great!  Is this ready for review now?

estebank commented:
Yes, it is 😃

solson commented:
I would change `bound in different ways` to `bound in a different way` on those span notes, since they are pointing at a single binding.

nikomatsakis commented:
@estebank quite nice

nikomatsakis commented:
@estebank I wonder if we should simplify the message to not talk about pattern numbers anymore, since we're showing it. i.e., instead of:

> variable `y` is bound with different mode in pattern #2 than in pattern #1

we can say

> inconsistent binding modes for variable `y` 

or 

> variable `y` is bound in inconsistent ways within the same match arm

(match arm is jargon, but maybe it's helpful to introduce it?)

nikomatsakis commented:
Similarly, something like

> variable `a` is not bound in all patterns

in place of

> variable `a` from pattern #2 is not bound in pattern #1

jonathandturner commented:
Yeah, this is definitely looking pretty good.  I'd be happy to approve if you're ready.

estebank commented:
Go ahead :)

jonathandturner commented:
@bors r+

bors commented:
:pushpin: Commit 01cb587 has been approved by `jonathandturner`

<!-- @bors r=jonathandturner 01cb5873acf4e193980243ea35be51e5c87469b2 -->

frewsxcv commented:
@bors r-

This failed on Travis https://travis-ci.org/rust-lang/rust/jobs/202465279

estebank commented:
From what I see I believe I'll have to sort the spans and diagnostics to make sure they are consistent throughout versions and platforms.

estebank commented:
@jonathandturner, @frewsxcv changed the test slightly to make the output deterministic across platforms. Should be ok to merge now.

jonathandturner commented:
@bors r+

bors commented:
:pushpin: Commit 6479044 has been approved by `jonathandturner`

<!-- @bors r=jonathandturner 6479044266ca97a7af2c65fa7228df064e7d6c6a -->

bors commented:
:hourglass: Testing commit 6479044 with merge ccea1bb...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/203661331)

estebank commented:
Error when downloading packages. Not caused by the patch.

jonathandturner commented:
@bors retry

bors commented:
:hourglass: Testing commit 6479044 with merge 09e19ac...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.2047)

jonathandturner commented:
@estebank - seems like the failure is potentially legit, as if the error still isn't quite deterministic

https://ci.appveyor.com/project/rust-lang/rust/build/1.0.2047/job/bys2pl5ekyaomw69

estebank commented:
@jonathandturner sorry for the slow turn around, was offline for the past two weeks.

The code [is now sorting all spans](https://github.com/rust-lang/rust/pull/39713/files#diff-aeb0880081a991f34aef2ab889e1fb7aR237), to avoid differences caused by different orderings when traversing HashMaps across platforms. I believe this _should_ be enough. While I was at it, I also rebased and squashed against last master.

eddyb commented:
@estebank You can always use `BTreeMap` instead of `HashMap`.

jonathandturner commented:
@estebank - can you sync with @eddyb to see if there's a solution that works with him? (eg his proposal of using `BTreeMap`)

estebank committed:
Use `BTreeSet` instead of `FxHashSet`

estebank commented:
@jonathandturner sure!

@eddyb I changed the [code to use `BTreeSet` instead of `FxHashSet`](https://github.com/rust-lang/rust/pull/39713/commits/780bea5067858737ebee373405698f0e4f9141f8). Does that seem reasonable?

jonathandturner commented:
Great!  Thanks @estebank 

@bors r+

bors commented:
:pushpin: Commit 780bea5 has been approved by `jonathandturner`

<!-- @bors r=jonathandturner 780bea5067858737ebee373405698f0e4f9141f8 -->

bors commented:
:lock: Merge conflict

estebank commented:
Rebased against master HEAD because of the bors merge conflict message, but [wasn't necessary](https://github.com/rust-lang/rust/pull/40285#issuecomment-284293198).

alexcrichton commented:
@bors: r=jonathandturner

bors commented:
:pushpin: Commit 3ffa4b5 has been approved by `jonathandturner`

<!-- @bors r=jonathandturner 3ffa4b52404d9a6aff7390d44336f85166235ff5 -->

bors commented:
:hourglass: Testing commit 3ffa4b5 with merge 2e6413c...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/208653831)

alexcrichton commented:
@bors: retry

* osx linker segfault #38878

On Tue, Mar 7, 2017 at 12:08 PM, bors <notifications@github.com> wrote:

> 💔 Test failed - status-travis
> <https://travis-ci.org/rust-lang/rust/builds/208653831>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/39713#issuecomment-284806572>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAD95AeIdXd99eLx7aikJGD-6E7M7ar6ks5rjZ0KgaJpZM4L86IO>
> .
>

bors commented:
:hourglass: Testing commit 3ffa4b5 with merge f91c3f6...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.2275), [status-travis](https://travis-ci.org/rust-lang/rust/builds/208903578)
Approved by: jonathandturner
Pushing f91c3f675517ed9676085d75d682f393b4108f96 to master...
